### PR TITLE
Bug 2104784: AWS: Fix race in IP address assignment code 

### PR DESCRIPTION
**Pull request by andreaskaris**
Signed-off-by: Andreas Karis <ak.karis@gmail.com>

FYI I also talked to Alex about this fix here and about why it was originally done differently; Alex shot for idempotency, but did not consider this specific race condition here.

### CODE REVIEW

File: pkg/cloudprovider/aws.go [hunk 1]
```diff
@@ -71,20 +72,34 @@ func (a *AWS) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
...
 	keepIPs := []*string{}
+	newIPs := []*string{}
 	if utilnet.IsIPv6(ip) {
 		for _, assignedIPv6 := range networkInterface.Ipv6Addresses {
 			if assignedIP := net.ParseIP(*assignedIPv6.Ipv6Address); assignedIP != nil && assignedIP.Equal(ip) {
 				return AlreadyExistingIPError
 			}
 			keepIPs = append(keepIPs, assignedIPv6.Ipv6Address)
 		}
-		keepIPs = append(keepIPs, &addIP)
```
Comment (kyrtapz) [inline pkg/cloudprovider/aws.go:81]: What's the point of having `keepIPs` now? we don't use it to assign the new IP anyway

File: pkg/cloudprovider/aws.go [hunk 2]
```diff
@@ -71,20 +72,34 @@ func (a *AWS) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
...
+		newIPs = append(newIPs, &addIP)
 		input := ec2.AssignIpv6AddressesInput{
 			NetworkInterfaceId: networkInterface.NetworkInterfaceId,
-			Ipv6Addresses:      keepIPs,
+			Ipv6Addresses:      newIPs,
```
Comment (kyrtapz) [inline pkg/cloudprovider/aws.go:86]: can we just create the array here and skip declaring `newIPs`?

File: pkg/cloudprovider/aws.go [hunk 3]
```diff
@@ -71,20 +72,34 @@ func (a *AWS) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
...
+				keepIPsStr = fmt.Sprintf("%s %s", keepIPsStr, *ip)
+			}
+			klog.Warningf("error: %s, tried to assign the following %d IP addresses to the interface:%v. Already got %d IP addresses on the interface: %v",
+				err.Error(),
+				len(newIPs), newIPsStr,
+				len(keepIPs), keepIPsStr,
+			)
 			return err
 		}
 		return a.waitForCompletion(node, awsapi.StringValueSlice(keepIPs), false)
```
Comment (kyrtapz) [inline pkg/cloudprovider/aws.go:105]: I think this should change to `return a.waitForCompletion(node, awsapi.StringValueSlice(newIPs), false)`

File: pkg/cloudprovider/aws.go (hunk 1 above)
Comment (andreaskaris) [reply pkg/cloudprovider/aws.go:81]: I renamed it to exsitingIPs and I added a comment above for why we need it. In go, we cant print the content of a slice of pointers, we'll get the only the pointers if we `fmt.Sprintf("%v", sliceOfPointers)`. So unless you know of something more elegant, we still must collect those IPs in a new []string{} slice

**Comment by andreaskaris** [comment]
The failing EgressIP test is a flake in the test itself. I ran the test manually multiple times against this very PR and it passes without issues.
~~~
akaris@linux origin ((2ca9c0cc41...))]$ TEST_SUITE_START_TIME="1658754962" KUBE_TEST_REPO_LIST="" KUBE_TEST_REPO="quay.io/openshift/community-e2e-images" TEST_PROVIDER="{\"type\":\"aws\",\"ProjectID\":\"\",\"Region\":\"us-west-2\",\"Zone\":\"us-west-2b\",\"NumNodes\":3,\"MultiMaster\":true,\"MultiZone\":true,\"Zones\":[\"us-west-2b\",\"us-west-2c\"],\"ConfigFile\":\"\",\"Disconnected\":false,\"SingleReplicaTopology\":false,\"NetworkPlugin\":\"OpenShiftSDN\",\"HasIPv4\":true,\"HasIPv6\":false,\"HasSCTP\":false,\"IsProxied\":false,\"IsIBMROKS\":false}" TEST_JUNIT_DIR="" TEST_UPGRADE_OPTIONS="" ./openshift-tests run-test "[sig-network][Feature:EgressIP] [external-targets] EgressIPs can be assigned automatically [Skipped:Network/OVNKubernetes] [Serial] [Suite:openshift/conformance/serial]"
~~~

**Pull request merged**